### DISCUSSION TITLE
Writing a postprocessor kernel to calculate the kinetic energy

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#25633] [Q&A General]
Hello everyone,

I am trying to write a new kernel similar to the LevelSetVolume. I want it to calculate the volume integral of the kinetic energy inside the c=0 contour. I am doing the following, but I see this error.
```

missing required parameter 'Postprocessors/Total_KE/vel_x'
	Doc String: "The x-axis velocity"
```

```
#pragma once

// MOOSE includes
#include "ElementVariablePostprocessor.h"

/**
 * Postprocessor to compute the area/volume inside and outside of a level set contour.
 */
class LevelSetVolumeKE : public ElementVariablePostprocessor
{
public:
  static InputParameters validParams();

  LevelSetVolumeKE(const InputParameters & parameters);

  virtual void initialize() override;
  virtual void finalize() override;
  virtual void execute() override;
  virtual Real getValue() override;
  virtual void threadJoin(const UserObject & y) override;
  virtual void computeQpValue() override{};

protected:
  /// The accumulated volume to return as a PostprocessorValue
  Real _volume;

  /// The level set contour to consider for computing inside vs. outside of the volume
  const Real & _threshold;

  /// Flag for triggering the internal volume calculation
  const bool _inside;
  const VariableValue & _u_vel;
  const VariableValue & _v_vel;
  const VariableValue & _w_vel;
};
```

```
#include "LevelSetVolumeKE.h"

registerMooseObject("LevelSetApp", LevelSetVolumeKE);

InputParameters
LevelSetVolumeKE::validParams()
{
  InputParameters params = ElementVariablePostprocessor::validParams();
  params.addClassDescription(
      "Compute the area or volume of the region inside or outside of a level set contour.");
  params.addParam<Real>(
      "threshold", 0.0, "The level set threshold to consider for computing area/volume.");

  MooseEnum loc("inside=0 outside=1", "inside");
  params.addParam<MooseEnum>("location", loc, "The location of the area/volume to be computed.");
  params.addRequiredCoupledVar("vel_x", "The x-axis velocity");
  params.addCoupledVar("vel_y", 0, "The y-axis velocity");
  params.addCoupledVar("vel_z", 0, "The z-axis velocity");
  return params;
}

LevelSetVolumeKE::LevelSetVolumeKE(const InputParameters & parameters)
  : ElementVariablePostprocessor(parameters),
    _threshold(getParam<Real>("threshold")),
    _inside(getParam<MooseEnum>("location") == "inside"),
    _u_vel(coupledValue("u")),
    _v_vel(_mesh.dimension() >= 2 ? coupledValue("v") : _zero),
    _w_vel(_mesh.dimension() == 3 ? coupledValue("w") : _zero)
{
}

void
LevelSetVolumeKE::initialize()
{
  _volume = 0;
}

void
LevelSetVolumeKE::execute()
{
  Real cnt = 0;
  Real n = _u.size();

  // Perform the check for inside/outside outside the qp loop for speed
  if (_inside)
  {
    for (_qp = 0; _qp < n; ++_qp)
      if (_u[_qp] <= _threshold)
        cnt++;
  }
  else
  {
    for (_qp = 0; _qp < n; ++_qp)
      if (_u[_qp] > _threshold)
        cnt++;
  }
  _volume += cnt / n * _current_elem_volume * 0.5 * (_u_vel[_qp]*_u_vel[_qp]+_v_vel[_qp]*_v_vel[_qp]+_w_vel[_qp]*_w_vel[_qp]);
}

void
LevelSetVolumeKE::finalize()
{
  gatherSum(_volume);
}

Real
LevelSetVolumeKE::getValue()
{
  return _volume;
}

void
LevelSetVolumeKE::threadJoin(const UserObject & y)
{
  const LevelSetVolumeKE & pps = static_cast<const LevelSetVolumeKE &>(y);
  _volume += pps._volume;
}
```

**Comment by GiudGiud** [comment]
Hello

Looks like you are adding a required coupled variable parameter to specify the X axis, but then in the input file this parameter is missing?

How are you using this object in the input file?

Guillaume

**Reply by alimostafavi24** [reply]
```
     [./Total_KE]
    type = LevelSetVolumeKE
    threshold = 0.0
    u = vel_x
    v = vel_y
    variable = c

    location = outside

    execute_on = 'initial timestep_end'
  [../]
```

**Reply by GiudGiud** [reply]
You called the parameter vel_x not u?

**Reply by alimostafavi24** [reply]
Yes, the main variable name in the input file is vel_x.

**Reply by GiudGiud** [reply]
Yes but you set the u parameter instead of the vel_x parameter

Try vel_x = vel_x for thé parameter

**Reply by alimostafavi24** [reply]
Thanks, it fixed the issue and now it is running.

However, it just returns zero means that postprocessor is not working correctly. I know that velocity field is correctly calculated. Is it the correct way to calculate the integral of  (0.5*velocity^2)d(volume)?

**Reply by GiudGiud** [reply]
No it doesn't look right. 
You re writing the computeIntegral function instead of just the Qp term, e.g. the computeQpIntegral function 

I would look at another postprocessor that does an integral for an example

**Reply by alimostafavi24** [reply]
ElementIntegralVariablePostprocessor is the right one which needs modification. It needs to get the phase field variable (c) and then perform integration only inside of the c=0 contour quadrature points, not the whole domain. Is this possible? I would highly appreciate it if I can get some help for this.

**Reply by GiudGiud** [reply]
where does c=0 happen? on the quadrature points in an element, or really anywhere within the elements?

**Reply by alimostafavi24** [reply]
I think the quadrature points since mesh is pretty fine in that region.

For example, I use levelsetvolume kernel to calculate the volume inside c=0 contour and it works very well.

Here some images to show the mesh and c=0 contour.

![Screenshot from 2023-09-30 14-04-37](https://github.com/idaholab/moose/assets/119982603/f87e182a-dfeb-4f72-82f9-59abc076969d)
![Screenshot from 2023-09-30 14-04-49](https://github.com/idaholab/moose/assets/119982603/930fbd82-38e0-4fb4-8eb4-6026936b9283)
![Screenshot from 2023-09-30 14-06-30](https://github.com/idaholab/moose/assets/119982603/c46fd0c8-0ae8-4301-8dd9-bc261e45ed8c)

**Reply by GiudGiud** [reply]
if it's just on the 'inside' of the contour, it's still a volume right?
could you just modify computeQpIntegral to do:

if (c[qp] > 0) 
  return 0;

so you only count when c <= 0

**Reply by alimostafavi24** [reply]
Thank you so much. It is working reasonably.
